samples-and-tests/just-test-cases/test/FunctionalTestTest.java

```diff
@@ -120,13 +121,70 @@ public void canGetRenderArgs() {
         User u = (User) renderArgs("u");
         assertEquals("Guillaume", u.name);
     }
-    
-    @Test
+
+    @Test(expected = RenderStatic.class)
```

Comment:
The changes to how exceptions are thrown back to tests has resulted in RenderStatic.class being thrown back up to the tests in this case. 

This change passes the test but I'm not 100% convinced this is the correct approach. Could someone who has more experience in how the static files are handled take a look at this please.

Reply:
@Fraserhardy @hsch Certainly, it's a bad idea to expect `RenderStatic.class` in this test. But it's not your fault.

There is two problems with this test:
1. it WAS incorrect before your change. It actually didn't test anything. If only you add a line `assertContentEquals("test", response);` to this test, it would fail.

2. Bad design of method `play.mvc.Router.Route#matches(java.lang.String, java.lang.String, java.lang.String, java.lang.String)`: it has size effect. Instead of returning `Map<String, String>`, it sometimes throws `RenderStatic`. And `PlayHandler` handles it as a special case: catches `RenderStatic` and sends file content asynchronously to the browser. Disgusting. :(

What I suggest is making a similar "special case" in `FunctionalTest`:

```java
        catch (ExecutionException e) {
          RuntimeException originalException = unwrapOriginalException(e);
          if (originalException instanceof RenderStatic) {
            response.status = 200;
            response.direct = ((RenderStatic) originalException).file;
          }
          else {
              throw originalException;
          }
        }
```

and adding a method in `FuncationalTest`:

```java
    public static void assertIsStaticFile(Response response, String filePath) {
        assertIsOk(response);

        String file = (String) response.direct;
        assertEquals(filePath, file);
    }
```

Finally, this test should be rewritten:

```java
    @Test
    public void testGettingStaticFile() {
	      Response response = GET("/public/session.test?req=1");
	      assertIsStaticFile(response, "public/session.test");
    }
```